core/vm/contracts_write.go

```diff
@@ -0,0 +1,192 @@
+func RunWriteCapablePrecompiledContract(
+	p WriteCapablePrecompiledContract,
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+	readOnly bool,
+) ([]byte, error) {
+	// immediately error out if readOnly
+	if readOnly {
+		return nil, errWriteProtection
+	}
+	gas, err := p.RequiredGas(evm, contract, input)
+	if err != nil {
+		return nil, err
+	}
+	if !contract.UseGas(gas) {
+		return nil, ErrOutOfGas
+	}
+	return p.RunWriteCapable(evm, contract, input)
+}
+
+type stakingPrecompile struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *stakingPrecompile) RequiredGas(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+) (uint64, error) {
+	// if invalid data or invalid shard
+	// set payload to blank and charge minimum gas
+	var payload []byte = make([]byte, 0)
+	// availability of staking and precompile has already been checked
+	if evm.Context.ShardID == shard.BeaconChainShardID {
+		// check that input is well formed
+		// meaning all the expected parameters are available
+		// and that we are only trying to perform staking tx
+		// on behalf of the correct entity
+		stakeMsg, err := staking.ParseStakeMsg(contract.Caller(), input)
+		if err == nil {
+			// otherwise charge similar to a regular staking tx
+			if encoded, err := rlp.EncodeToBytes(stakeMsg); err == nil {
+				payload = encoded
+			}
+		}
+	}
+	if gas, err := IntrinsicGas(
+		payload,
+		false,                                   // contractCreation
+		evm.ChainConfig().IsS3(evm.EpochNumber), // homestead
+		evm.ChainConfig().IsIstanbul(evm.EpochNumber), // istanbul
+		false, // isValidatorCreation
+	); err != nil {
+		return 0, err // ErrOutOfGas occurs when gas payable > uint64
+	} else {
+		return gas, nil
+	}
+}
+
+// RunWriteCapable runs the actual contract (that is it performs the staking)
+func (c *stakingPrecompile) RunWriteCapable(
+	evm *EVM,
+	contract *Contract,
+	input []byte,
+) ([]byte, error) {
+	if evm.Context.ShardID != shard.BeaconChainShardID {
+		return nil, errors.New("Staking not supported on this shard")
+	}
+	stakeMsg, err := staking.ParseStakeMsg(contract.Caller(), input)
+	if err != nil {
+		return nil, err
+	}
+	if delegate, ok := stakeMsg.(*stakingTypes.Delegate); ok {
+		if err := evm.Delegate(evm.StateDB, delegate); err != nil {
+			return nil, err
+		} else {
+			evm.StakeMsgs = append(evm.StakeMsgs, delegate)
+			return nil, nil
+		}
+	}
+	if undelegate, ok := stakeMsg.(*stakingTypes.Undelegate); ok {
+		return nil, evm.Undelegate(evm.StateDB, undelegate)
+	}
+	if collectRewards, ok := stakeMsg.(*stakingTypes.CollectRewards); ok {
+		return nil, evm.CollectRewards(evm.StateDB, collectRewards)
+	}
+	return nil, stakingTypes.ErrInvalidStakingKind
+}
+
+type migrationPrecompile struct{}
```

Comment:
Why not make Migration part of the stakingPrecompile as an option?  No need to create a new precompile for it imo. It's part of the staking precompile logic.

Reply:
Done.